Remove web framework dependency from tools

davelopez opened the pull request:
Part of #17046

This tries to remove several imports of `galaxy.web` from `galaxy.tools` and `galaxy.managers`.
Most changes involve handling URL generation without directly importing `galaxy.web.url_for`.

## How to test the changes?
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

davelopez committed:
Remove web dependency from tools/errors.py

davelopez committed:
Remove web dependency from tools/error_reports

davelopez committed:
Remove web.framework dependency from tools

- Use `util.unicodify` in copy_view.mako for consistency with other usages of unicodify in the file.
- Move `to_unicode` to the only place it's used under tool/search/__init__.py

davelopez committed:
Move short_term_storage related classes to managers

The ShortTermStorageManager is a manager and makes sense to have it live under galaxy.managers instead of galaxy.web.

It will also help reduce the dependency of the manager modules on web related stuff.

davelopez committed:
Move api_error_message from web decorators to exceptions.utils

- And rename it to `api_error_to_dict` since it creates dictionary with and not a message string as the previous name might suggest.

nsoranzo commented on lib/galaxy/exceptions/utils.py:
```diff
@@ -0,0 +1,32 @@
+from galaxy.exceptions import error_codes
+
+
+def api_error_to_dict(trans, **kwds):
```
```suggestion
def api_error_to_dict(**kwds):
```

nsoranzo commented on lib/galaxy/exceptions/utils.py:
```diff
@@ -0,0 +1,32 @@
+    # err_msg used a good number of places already. Might as well not change
+    # it?
+    error_response = dict(err_msg=err_msg, err_code=error_code, **extra_error_info)
+    if trans and trans.debug:  # TODO: Should admins get to see traceback as well?
```
```suggestion
    if kwds.get("debug"):  # TODO: Should admins get to see traceback as well?
```

nsoranzo commented on lib/galaxy/web/framework/decorators.py:
```diff
@@ -395,39 +396,8 @@ def validation_error_to_message_exception(e: ValidationError) -> MessageExceptio
-
 def __api_error_dict(trans, **kwds):
-    error_dict = api_error_message(trans, **kwds)
+    error_dict = api_error_to_dict(trans, **kwds)
```
```suggestion
    error_dict = api_error_to_dict(debug=trans.debug, **kwds)
```

nsoranzo replied:
Ideally this should go into the "[Remove unnecessary trans parameter](https://github.com/galaxyproject/galaxy/pull/17058/commits/80d06bc637c8366e58da725db4ed07902303d083)" commit :wink:

jmchilton commented on lib/galaxy/app.py:
```diff
@@ -52,6 +52,12 @@
+    ShortTermStorageConfiguration,
+    ShortTermStorageManager,
+    ShortTermStorageMonitor,
+)
```
I dislike this a lot - managers bring in so many dependencies. galaxy.web.short_term_storage was a nice little isolated microservice effectively and now it just mixed in with the rest of galaxy-app.

davelopez replied:
Then I'm happy you caught this on time :)
I moved it to managers because the Manager suffix in ShortTermStorageManager suggested me to do so, but we can put it where it makes more sense. Where do you think should be the best fitting place?

davelopez replied:
This discussion continues here https://github.com/galaxyproject/galaxy/pull/17058#discussion_r1400909897

jmchilton commented on lib/galaxy/exceptions/utils.py:
```diff
@@ -0,0 +1,32 @@
+from galaxy.exceptions import error_codes
+
+
+def api_error_to_dict(trans, **kwds):
```
I do on the other hand, really like this. This is a nice enhancement IMO.

davelopez replied:
Marius' idea, it has to be ;)

jmchilton commented on lib/galaxy/managers/short_term_storage.py:
```diff
@@ -24,13 +24,13 @@
     is_uuid,
     safe_makedirs,
 )
-from galaxy.web.framework.decorators import api_error_message
```
Could STS be placed in its own package now with this change? All it depends on now is galaxy-util and galaxy schema stuff (and then only for the typing).

davelopez replied:
Yes sure! Do you mean a whole new package? or do you have an idea where to put it?

davelopez replied:
galaxy.microservices? :thinking:

jmchilton replied:
I think the idea of a microservices package runs counter the idea of doing microservices 😅. I think the path should maybe just be ``galaxy.sts`` or ``galaxy.short_term_storage``. As to what Python package that module belongs to... that is trickier. 

I would think that should be in its own package based on its dependencies. If that is too heavy moving it to galaxy.web_stack would be possible after this commit since you've refactored the galaxy.web.framework pieces out. It doesn't belong there conceptually but it is a lighter weight import than all the web framework stuff. You could also keep it in web framework for now and just add a TODO somewhere that when we manage to decompose tools from the web framework we need to spin STS out.

davelopez replied:
> I think the idea of a microservices package runs counter the idea of doing microservices 😅

Hah true! I see your point now :laughing:

Ok, I will see what I can do and in the worst case we do the TODO as suggested :+1:

jmchilton commented on lib/galaxy/tools/error_reports/plugins/slack.py:
```diff
@@ -38,7 +37,7 @@ def _fill_issue_cache(self, *args, **kwargs):
     def submit_report(self, dataset, job, tool, **kwargs):
         history_id_encoded = self.app.security.encode_id(dataset.history_id)
-        history_view_link = web.url_for("/histories/view", id=history_id_encoded, qualified=True)
+        history_view_link = self.app.url_for("/histories/view", id=history_id_encoded, qualified=True)
```
+1

nsoranzo commented on lib/galaxy/tools/search/__init__.py:
```diff
@@ -57,15 +57,25 @@
 CanConvertToInt = Union[str, int, float]
 
 
+def to_unicode(a_string):
```
I am pretty sure there is no need for this function because:
- it could be replaced with just `unicodify()` as it misuses the encoding parameter ("utf-8"), which is the source encoding, not the desired encoding.
- all calls to this functions are not needed any way, as it's used on `Tool` attributes which are strings.

davelopez replied:
Thanks! I will remove it then, I just did the minimal move here but it makes sense if it was not that useful in the first place.

davelopez committed:
Remove unnecessary trans parameter

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

davelopez committed:
Replace to_unicode with unicodify

Co-authored-by: Nicola Soranzo <nicola.soranzo@earlham.ac.uk>

davelopez committed:
Move short_term_storage related classes to its own path

The ShortTermStorageManager and related classes should live in their own package eventually and not within galaxy.web nor galaxy.managers as the short term storage microservices it's fairly independent and has way less dependencies than those other packages.

nsoranzo commented on lib/galaxy/webapps/galaxy/api/short_term_storage.py:
```diff
@@ -1,9 +1,9 @@
 """
-API operations around galaxy.web.short_term_storage infrastructure.
+API operations around galaxy.managers.short_term_storage infrastructure.
```
```suggestion
API operations around galaxy.short_term_storage infrastructure.
```

davelopez committed:
Use correct module name in comment

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

davelopez committed:
Add symlink to short_term_storage in app package

Temporarily include it in app until the STS goes into its own package.

nsoranzo commented on test/unit/web_framework/test_short_term_storage.py:
```diff
@@ -7,7 +7,7 @@
     MessageException,
```
If we agree to move `short_term_storage` to the `app` package for the time being, then also this file should be moved to `test/unit/app/`

davelopez replied:
Thanks! I will move it for now to test/unit/app/

davelopez committed:
Move STS unit tests to test/unit/app

Temporarily until we create its own package.

davelopez committed:
Remove galaxy.web import from managers.base

We need to use url_builder from the request context to be able to generate valid URLs.

Hopefully url_for_not_available will help identify any other possible use of url_for that does not fall back into url_builder.

davelopez committed:
Remove galaxy.web import from managers.workflows

Replace calls to web.url_for with compatible url_builder.

davelopez committed:
Remove galaxy.web import from managers.users

Replace calls to web.url_for with compatible url_builder.

Since there is no real API route for those (they live in client controllers) we need to pass the path prefix as name...

nsoranzo commented on lib/galaxy/managers/workflows.py:
```diff
@@ -1584,12 +1587,12 @@ def callback(input, prefixed_name, **kwargs):
         return data
 
-    def _workflow_to_dict_instance(self, stored, workflow, legacy=True):
+    def _workflow_to_dict_instance(self, stored, workflow, legacy=True, url_builder=None):
```
Maybe have `trans` as first parameter instead, as the other similar methods?

davelopez replied:
I tried to explicitly avoid passing `trans` and be more specific about what this function really needs from it, but I guess it's not worth the inconsistency with the rest of the functions, and the ship of "not using trans everywhere" has sailed already :laughing: 

I will do it for consistency.

nsoranzo commented on lib/galaxy/managers/users.py:
```diff
@@ -499,8 +498,11 @@ def send_activation_email(self, trans, email, username):
-        activation_link = url_for(
-            controller="user", action="activate", activation_token=activation_token, email=escape(email), qualified=True
+        activation_link = trans.url_builder(
+            "/activate",
```
Should this be:
```suggestion
            "/user/activate",
```
?

davelopez replied:
Surprisingly both links work:
- `http://127.0.0.1:8081/activate?activation_token={token}&email={email}`
- `http://127.0.0.1:8081/user/activate?activation_token={token}&email={email}`

The previous `url_for` was generating the one without `/user`, but if both work I think is more meaningful to generate the one with `/user` as you suggested :+1:

davelopez committed:
Change account activation link to /user/activate

Both links `{host}/activate` and `{host}/user/activate` work but the latter seems more explicit about what are we trying to activate.

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

davelopez committed:
Remove galaxy.web import from managers.configuration

Taking the server_starttime from app instead.

davelopez committed:
Pass trans to _workflow_to_dict_instance

Mostly for consistency with other similar functions in the module.

Co-authored-by: Nicola Soranzo <nicola.soranzo@earlham.ac.uk>

davelopez committed:
Adapt unit tests to use url_builder mock

davelopez committed:
Pass missing context when serializing urls

This should fix URL generation for collections by using the URL builder instead of the legacy url_for.

nsoranzo commented on test/unit/app/managers/test_HDCAManager.py:
```diff
@@ -58,7 +58,7 @@ def testable_url_for(*a, **k):
     return f"(fake url): {a}, {k}"
```
You can replace `testable_url_for` with `from galaxy.app_unittest_utils.galaxy_mock import mock_url_builder`

davelopez committed:
Replace testable_url_for with mock_url_builder

nsoranzo commented on lib/galaxy/app.py:
```diff
@@ -800,7 +800,7 @@ def __init__(self, **kwargs) -> None:
         self.legacy_url_for = legacy_url_for
 
-        self.server_starttime = server_starttime  # used for cachebusting
+        self.start_timestamp = self.server_starttime = server_starttime  # used for cachebusting
```
Do we need 2 attributes for this? Not sure if `self.start_timestamp` is needed or we could just use `self.server_starttime` ?

davelopez replied:
We don't really need 2, but I created an alias because I was concerned about the naming containing "server" on it as it will be exposed now in the `StructuredApp` interface (that is supposed to not have anything web-related on it but I can see now `api_keys_manager` is on it too...).

I also don't like that duplication, but I guess it makes sense to call it "server_starttime" in some contexts and "start_timestamp" (or even better "app_starttime"?) in others... or we can just choose one and rename all usages. What do you prefer?

nsoranzo replied:
I don't think server strictly implies a *web* server, and I find "starttime" clearer than "timestamp" here. Also, we use `server_startttime` as part of the `/api/configuration` output.
But as long there is one blessed name, I don't mind particularly.

davelopez replied:
Alright! Thanks for the suggestion, I changed it back to `server_startttime` :+1:

davelopez committed:
Revert the use of the  walrus := operator

This package is used by pulsar with still supports Python 3.7

davelopez committed:
Revert renaming of server_starttime
